frame/balances/src/lib.rs

```diff
@@ -1656,23 +1646,47 @@ where
 
 	type MaxLocks = T::MaxLocks;
 
+	// Checks if `set_lock` or `extend_lock` would fail.
+	// Specifically if adding a lock with `id` would push us
+	// over the `MaxLocks` limit.
+	fn can_add_lock(id: LockIdentifier, who: &T::AccountId) -> bool {
```

Comment:
this function is correct, but a bit unneccesarily hard IMO. for example allocating via `.to_vec`  just to check a count can be prevented. 

Also, for the earlier logic, can do something like:

```rust
let locks = Self::locks(who);
let length = locks.len();
let new_locks_len =  if Self::locks(who).any(|l| l.id == id) { length } else { length + 1 }; 
```